// Copyright (C) 2020 Ellen Poe
//
// This file is part of MASS.
//
// MASS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// MASS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with MASS.  If not, see <http://www.gnu.org/licenses/>.

syntax = "proto3";
package api;

import "mass/api/chat.proto";
import "mass/api/spatial.proto";

message VesselUpdate {
  // Repeated field containing updates to various vessel systems.
  repeated SystemUpdate system_updates = 1;

  // Any new chat messages.
  repeated ChatMessage chat_updates = 2;
}

message SystemUpdate {
  oneof update {
    // Steering update.
    SteeringSystemUpdate steering_update = 1;
    // Diving update.
    DivingSystemUpdate diving_update = 2;
    // Propulsion update.
    PropulsionSystemUpdate propulsion_update = 3;
    // Map update.
    MapSystemUpdate map_update = 4;
    // Hull system update.
    HullSystemUpdate hull_update = 5;
  }
}

message SteeringSystemUpdate {
  // Requested heading.
  uint32 requested_heading_degrees = 1;

  // Actual heading.
  uint32 actual_heading_degrees = 2;
}

message DivingSystemUpdate {
  // Requested keel depth in feet.
  double requested_depth_feet = 1;
}

message PropulsionSystemUpdate {
  // Requested speed.
  uint32 requested_speed_knots = 1;

  // Actual speed.
  uint32 actual_speed_knots = 2;
}

message MapSystemUpdate {
  // The current position.
  Position position = 1;
}

message HullSystemUpdate {
  // Actual depth/draft in feet.
  double actual_depth_feet = 1;
}
